Automattic/sensei
# Use webpack for all JS and CSS builds

Compile all frontend assets with `calypso-build`.

### Goals

- Minimize per-project build tooling and configuration to reduce maintenance work
- Use built tooling from Wordpress/Gutenberg project if possible
- Add ability to use the React-based components from Wordpress and Woo component libraries
- Be able to add components gradually without having to rewrite the existing JS code
- Remove generated CSS and JS code from source tree, as these lead to extra work/mental burden and potential merge conflicts
- Keep it easy to work with code in various local/testing environments
- Possibly take advantage of webpack-dev-server in-memory compilation, live reload & hot module replacement opportunities for fast iteration

### Changes

This has two major parts, changing gulp build to webpack, and a PHP wrapper around `wp_enqueue_*` calls. 

- JS and SCSS build is now done in one step by `@automattic/calypso-build`, both for React/JSX files and legacy scripts.

- All generated files are in `assets/dist`, instead of `[filename].min.js`. This is easily excluded from git and IDEs, and is more in line with the default settings of the build system. 

- Remove `gulp` JS and CSS build, and the file watching functionality related to this. This is handled by webpack.

- Remove unused dependencies from `package.json`

- Use `wp-scripts` to call various commands like linting and unit testing. This runs the same things as previously, but it should include all Wordpress specific presets and configurations, so we don't have to maintain those in the plugins. 

### Changes in developer workflow: 

- **Removing generated code from the source:**
   This is not strictly necessary, but would likely result in a cleaner commit & PR workflow. The price is, contributors will have to run `npm install && npm run build` to get the frontend assets built.
   Linking or mounting the plugin folder to a Wordpress install still works.

- **Option to load uncompiled files via SCRIPT_DEBUG removed:**
   This wouldn't work for JSX files and files using module imports anyway. Compiled files in dev mode include source maps instead, so the original source can still be debugged in DevTools / IDE.

- **`wp_enqueue_scripts|styles` wrapper:**
   Loading scripts and styles should be done through the <code>Sensei_Assets</code> helper class, which handles loading the built assets and any dependencies they have. </p>

- **Cache busting should work:**
  The above helper will use the content hash of the generated assets as the file version, so the browser will always load a fresh version when a file changes.

### Testing instructions:

* After `npm install` to update, `npm run build` runs the build, and creates the compiled files in `assets/dist`.
* `npm run package` should produce a package zip in the `build` folder, that is working correctly.
* `npm run start` should build assets and watch for changes, and produce a development version from everything. In browser devtools, the original source files should be debuggable. 

* All custom scripts, select2 boxes and styles on the following pages should be loaded and working:
 - [x] Learner management
 - [x] Email signup modal
 - [x] Course edit page - prerequisites
 - [x] Frontend: Tabs on /learner/username
 - [x] Lesson, quiz select boxes
 - [x] Order lessons, modules, courses
 - [x] Grading
 - [x] Settings, range slider
 - [x] Lessons bulk edit

### Further tasks to be tackled in additional PRs:
* Use this in plugins, especially WCPC. 
* Some optimizations in output file generation.
* Improve handling some vendor files like select2 and jquery-ui bits.
* Optimize bundling Woo components
* Remove unused `chosen` library
